src/equations/shallow_water_exner_1d.jl

```diff
@@ -0,0 +1,618 @@
+The equations are given by
+```math
+\begin{cases}
+\partial_t h + \partial_x hv = 0, \\
+\partial_t hv + \partial_x (hv^2) + gh\partial_x (h + h_b) + g\frac{1}{r}h_s\partial_x (rh + h_b) + \frac{\tau}{\rho_f} = 0,\\
+\partial_t h_b + \partial_x q_s = 0,
+\end{cases}
+```
+The unknown quantities are the water and sediment height ``h``, ``h_b`` and the velocity ``v``.
+The sediment discharge ``q_s(h, hv)`` is determined by the `sediment_model` and is used to determine 
+the active sediment height ``h_s = q_s / v``. 
+Furthermore ``\tau`` denotes the shear stress at the water-sediment interface and is determined by 
+the `friction` model.  
+The gravitational constant is denoted by ``g``, and ``\rho_f`` and ``\rho_s`` are the fluid and sediment
+densities, respectively. The density ratio is given by ``r = \rho_f / \rho_s``.
+
+The conservative variable water height ``h`` is measured from the sediment height ``h_b``, therefore
+one also defines the total water height as ``H = h + h_b``.
+
+The additional quantity ``H_0`` is also available to store a reference value for the total water 
+height that is useful to set initial conditions or test the "lake-at-rest" well-balancedness.
+
+The entropy conservative formulation has been derived in the paper:
+- E.D. Fernández-Nieto, T.M. de Luna, G. Narbona-Reina and J. de Dieu Zabsonré (2017)\
+  Formal deduction of the Saint-Venant–Exner model including arbitrarily sloping sediment beds and 
+  associated energy\
+  [DOI: 10.1051/m2an/2016018](https://doi.org/10.1051/m2an/2016018)
+"""
+struct ShallowWaterExnerEquations1D{RealT <: Real,
+                                    FrictionT <: Friction{RealT},
+                                    SedimentT <: SedimentModel{RealT}} <:
+       Trixi.AbstractShallowWaterEquations{1, 3}
+    # This structure ensures that friction and sediment models are concrete types 
+    # to prevent allocations
+    gravity::RealT # gravitational constant
+    H0::RealT      # constant "lake-at-rest" total water height
+    friction::FrictionT
+    sediment_model::SedimentT
+    porosity_inv::RealT  # 1/(1-porosity)
+    rho_f::RealT    # density of fluid
+    rho_s::RealT    # density of sediment
+    r::RealT       # density ratio
+end
+
+# Allow for flexibility to set the gravitational constant within an elixir depending on the
+# application where `gravity_constant=1.0` or `gravity_constant=9.81` are common values.
+# The reference total water height H0 defaults to 0.0 but is used for the "lake-at-rest"
+# well-balancedness test cases.
+function ShallowWaterExnerEquations1D(; gravity_constant, H0 = zero(gravity_constant),
+                                      friction = ManningFriction(n = 0.0),
+                                      sediment_model,
+                                      porosity, rho_f, rho_s)
+    # Precompute common expressions for the porosity and density ratio
+    porosity_inv = 1 / (1 - porosity)
```

Comment:
I think type stability shouldn't be a problem, since we specify that `porosity_inv::RealT`.
The result of `inv(1-porosity)` will have the same type as `porosity`. So as long as `porosity` is specified in the correct type everything should be fine.